docs/testing-strategy.md

```diff
@@ -0,0 +1,152 @@
+- pragmatism in what we test - if all code paths of a class/module are
+  exhausted by other tests then we may not need to explicitly test something
+  further.
+
+This testing strategy was inspired by:
+[Whitehall Testing Guidelines][whitehall-tests],
+[Email Alert API Testing][email-alert-api-tests], and
+[Thoughtbots' How We Test Rails Applications][thoughtbot-rails-tests]
+
+## Test organisation
+
+Tests for Content Publisher are stored in the [spec](../spec) directory with
+the Ruby tests written using [RSpec](https://rspec.info/) and JavaScript tests
+written using [Jasmine](https://jasmine.github.io/).
+
+As per common Rails conventions, most of the directories within the spec
+directory contain Ruby [unit tests](#unit-tests) with the directories matching
+those within the app directory. Code that is stored in the lib directory has
+corresponding tests in the spec/lib directory.
+
+Directories within spec that don't contain Ruby unit tests are as follows:
+
+- factories - contains [FactoryBot factories][factory-bot], which are a fixture
+  alternative;
+- features - contains [feature tests](#feature-tests), which exercises
+  application features from a user perspective;
+- javascripts - contains [unit tests](#unit-tests) for JavaScript files
+  and configuration and helper files for Jasmine;
+- requests - contains [request tests](#request-tests), which test the HTTP API
+  of the application;
+- support - contains helper files for shared test methods and logic;
+- views - contains [view tests](#view-tests), which are used to test
+  [Rails views][rails-views] output expected HTML in a given scenario.
+
+## Unit tests
+
+The purpose of these tests is to test individual units of the system in
+isolation. They are intended to provide exhaustive tests of the code paths
+through a particular class or module through its public interface. As per
+the [test pyramid](https://martinfowler.com/bliki/TestPyramid.html) approach
+to software testing, these should provide the majority of test coverage.
+
+Characteristics of unit tests:
+
+- they should be concerned with the logic within the class/module being tested,
+  and not test logic defined elsewhere;
+- they may mock dependent objects and/or assert that particular external
+  methods are called;
+- they should test all code paths through a class/module;
+- they should test that side effects intended by the code occur, for example writing
+  to the database or making API calls.
+
+## Feature tests
+
+Feature tests are used to assert that Content Publisher behaves correctly from
+the perspective of someone using the application. This ensures we're testing
+the system in a way representative of expected user behaviour. This makes the
+tests some of the most valuable of the application, however they are slower
+than other tests to run and can be difficult to debug. Therefore they are not
+intended to exhaustively test all the scenarios that can occur.
```

Comment:
Hopefully this should be covered with: https://github.com/alphagov/content-publisher/blame/b55b02737011b70e64629e2090964458c3b96158/docs/testing-strategy.md#L86-L88

I wanted to avoid the phrase "happy path" as I think it's a bit more nuanced than that in that if you have a long negative flow that is a feature in itself like scheduled publishing failed it doesn't quite make sense.

Reply:
Ah, you have a point. 👍